File: 763-special-binary-string/special-binary-string.js
/**
 * @param {string} s
 * @return {string}
 */
var makeLargestSpecial = function(s) {
    var ans = "";
    let start = 0;
    let l = s.length;
    let sum = 0;
    let strs = [];
    for(let i = 0; i < l; i++){
        if(s[i] === '1'){
            sum++;
        }else{
            sum--;
        }

        if(sum === 0){
            //now we got a special string, let's see if there is any such string inside this
            var curMidMax = "";
            if(start+1 < i){
                curMidMax = makeLargestSpecial(s.slice(start+1, i));
            }
            var curMax = s[start];
            curMax += curMidMax;
            curMax += s[i];
            strs.push(curMax);
            start = i+1;
        }
    }
        strs.sort((a,b) => b.localeCompare(a));
        return strs.join("");
};

File: 763-special-binary-string/special-binary-string.cpp
class Solution {
public:
    string makeLargestSpecial(string s) {
        int start = 0;
        string ans = "";
        int l = s.length();
        vector<string> substrs;
        int sum = 0;
        for(int i = 0; i < l; i++){
            if(s[i] == '1') sum++;
            else sum--;

            if(sum == 0){
                string curMidMax = "";
                if(start < i){
                    string subs = s.substr(start+1, i - start - 1);
                    curMidMax = makeLargestSpecial(subs);
                }
                string curMax = "";
                curMax.push_back(s[start]);
                curMax += curMidMax;
                curMax.push_back(s[i]);
                substrs.push_back(curMax);
                start = i+1;
            }
            sort(substrs.begin(), substrs.end(), greater<string>());
        }
        for(string x : substrs) ans += x;
        return ans;
    }
};